crates/common/src/raindex_client/vaults.rs

```diff
@@ -132,12 +155,520 @@ impl RaindexVaultToken {
+    ///   console.error("Approval error:", result.error.readableMsg);
+    ///   return;
+    /// }
+    /// const calldata = result.value;
+    /// // Do something with the calldata
+    /// ```
+    #[wasm_export(js_name = "getApprovalCalldata", unchecked_return_type = "Hex")]
+    pub async fn get_approval_calldata(
+        &self,
+        deposit_amount: String,
+    ) -> Result<Bytes, RaindexError> {
+        let deposit_amount = self.validate_amount(&deposit_amount)?;
+
+        let (deposit_args, transaction_args) =
+            self.get_deposit_and_transaction_args(deposit_amount)?;
+
+        let allowance = deposit_args
+            .read_allowance(self.owner, transaction_args.clone())
+            .await?;
+        if allowance >= deposit_amount {
+            return Err(RaindexError::ExistingAllowance);
+        }
+
+        Ok(Bytes::copy_from_slice(
+            &deposit_args.get_approve_calldata(transaction_args).await?,
+        ))
+    }
+
+    /// Gets the current ERC20 allowance for a vault.
+    ///
+    /// Determines how much the orderbook contract is currently approved to spend
+    /// on behalf of the vault owner.
+    ///
+    /// ## Returns
+    ///
+    /// * `bigint` - Current allowance amount
+    ///
+    /// ## Examples
+    ///
+    /// ```javascript
+    /// const result = await vault.getAllowance();
+    /// if (result.error) {
+    ///   console.error("Cannot check allowance:", result.error.readableMsg);
+    ///   return;
+    /// }
+    /// const allowance = result.value;
+    /// // Do something with the allowance
+    /// ```
+    #[wasm_export(js_name = "getAllowance")]
+    pub async fn get_allowance(&self) -> Result<RaindexVaultAllowance, RaindexError> {
+        let (deposit_args, transaction_args) = self.get_deposit_and_transaction_args(U256::ZERO)?;
+        let allowance = deposit_args
+            .read_allowance(self.owner, transaction_args.clone())
+            .await?;
+        Ok(RaindexVaultAllowance(allowance))
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+#[wasm_bindgen]
+pub struct RaindexVaultBalanceChange {
+    __typename: String,
+    amount: U256,
+    new_vault_balance: U256,
+    old_vault_balance: U256,
+    timestamp: U256,
+    transaction: RaindexTransaction,
+    orderbook: Address,
+}
+#[wasm_bindgen]
+impl RaindexVaultBalanceChange {
+    #[wasm_bindgen(getter)]
+    pub fn __typename(&self) -> String {
+        self.__typename.clone()
+    }
+    #[cfg(target_family = "wasm")]
+    #[wasm_bindgen(getter)]
+    pub fn amount(&self) -> Result<BigInt, RaindexError> {
+        BigInt::from_str(&self.amount.to_string())
+            .map_err(|e| RaindexError::JsError(e.to_string().into()))
+    }
+    #[cfg(not(target_family = "wasm"))]
+    #[wasm_bindgen(getter)]
+    pub fn amount(&self) -> String {
+        self.amount.to_string()
+    }
+    #[cfg(target_family = "wasm")]
+    #[wasm_bindgen(getter = newVaultBalance)]
```

Comment:
is a getter the right approach if it's a Result?

would it be better to do all these fallible operations when originally creating the type so we don't have to do it later?